File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/component/PropertyBasedPermissionEvaluator.java
package com.windea.demo.cloudcollect.core.component;

import com.windea.demo.cloudcollect.core.domain.entity.*;
import com.windea.demo.cloudcollect.core.service.*;
import org.springframework.security.access.PermissionEvaluator;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;

import java.io.Serializable;
import java.util.Objects;
import java.util.Set;

/**
 * 自定义的许可鉴别器。基于实体类属性。判断此属性是否等于principal.name，且对应的permission是否相匹配。
 * <p>用于实现Spring El中的hasPermission()方法。
 * <p>TODO 加入缓存控制，提高扩展性。
 */
@Component
public class PropertyBasedPermissionEvaluator implements PermissionEvaluator {
	public final static Set<String> allPermissionNames = Set.of("read", "write", "create", "delete", "admin");
	public final static String targetDomainPackage = "com.windea.demo.cloudcollect.core.domain.entity";

	private final CollectService collectService;
	private final CollectCategoryService categoryService;
	private final CollectTagService tagService;
	private final CommentService commentService;
	private final NoticeService noticeService;

	public PropertyBasedPermissionEvaluator(CollectService collectService, CollectCategoryService categoryService,
		CollectTagService tagService, CommentService commentService, NoticeService noticeService) {
		this.collectService = collectService;
		this.categoryService = categoryService;
		this.tagService = tagService;
		this.commentService = commentService;
		this.noticeService = noticeService;
	}


	@Override
	public boolean hasPermission(Authentication authentication, Object targetDomainObject, Object permission) {
		var principal = "";
		var permissions = Set.of();

		if(targetDomainObject instanceof Collect) {
			principal = ((Collect) targetDomainObject).getUser().getUsername();
			permissions = Set.of("read", "write", "create", "delete");
		} else if(targetDomainObject instanceof CollectTag) {
			principal = ((CollectTag) targetDomainObject).getUser().getUsername();
			permissions = Set.of("read", "write", "create", "delete");
		} else if(targetDomainObject instanceof CollectCategory) {
			principal = ((CollectCategory) targetDomainObject).getUser().getUsername();
			permissions = Set.of("read", "write", "create", "delete");
		} else if(targetDomainObject instanceof Comment) {
			principal = ((Comment) targetDomainObject).getSponsorByUser().getUsername();
			permissions = Set.of("read", "create", "delete");
		} else if(targetDomainObject instanceof Notice) {
			principal = ((Notice) targetDomainObject).getUser().getUsername();
			permissions = Set.of("read", "delete");
		} else {
			return false;
		}

		return Objects.equals(authentication.getName(), principal) && permissions.contains(permission);
	}

	@Override
	public boolean hasPermission(Authentication authentication, Serializable targetId, String targetType,
		Object permission) {
		Object targetDomainObject;

		if(Objects.equals(targetType, "Collect")) {
			targetDomainObject = collectService.get((Long) targetId);
		} else if(Objects.equals(targetType, "CollectCategory")) {
			targetDomainObject = categoryService.get((Long) targetId);
		} else if(Objects.equals(targetType, "CollectTag")) {
			targetDomainObject = tagService.get((Long) targetId);
		} else if(Objects.equals(targetType, "Comment")) {
			targetDomainObject = commentService.get((Long) targetId);
		} else if(Objects.equals(targetType, "Notice")) {
			targetDomainObject = noticeService.get((Long) targetId);
		} else {
			return false;
		}

		return hasPermission(authentication, targetDomainObject, permission);
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/component/PropertyBasedPermissionEvaluator.kt
package com.windea.demo.cloudcollect.core.component

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.service.*
import org.springframework.security.access.*
import org.springframework.security.core.*
import org.springframework.stereotype.*
import java.io.*

/**
 * 自定义的许可鉴别器。基于实体类属性。判断此属性是否等于principal.name，且对应的permission是否相匹配。
 *
 * 用于实现Spring El中的hasPermission()方法。
 *
 * TODO 加入缓存控制，提高扩展性。
 */
@Component
class PropertyBasedPermissionEvaluator(
	private val collectService: CollectService,
	private val categoryService: CollectCategoryService,
	private val tagService: CollectTagService,
	private val commentService: CommentService,
	private val noticeService: NoticeService
) : PermissionEvaluator {
	override fun hasPermission(authentication: Authentication, targetDomainObject: Any, permission: Any): Boolean {
		val permissions: Set<String> = when(targetDomainObject) {
			is Collect -> setOf("read", "write", "create", "delete")
			is CollectTag -> setOf("read", "write", "create", "delete")
			is CollectCategory -> setOf("read", "write", "create", "delete")
			is Comment -> setOf("read", "create", "delete")
			is Notice -> setOf("read", "create", "delete")
			else -> return false
		}
		
		if(permission !in permissions) return false
		if(permission == "create") return true
		
		val principalName: String = when(targetDomainObject) {
			is Collect -> targetDomainObject.user.username
			is CollectTag -> targetDomainObject.user.username
			is CollectCategory -> targetDomainObject.user.username
			is Comment -> targetDomainObject.sponsorByUser.username
			is Notice -> targetDomainObject.user.username
			else -> return false
		}
		
		return principalName == authentication.name
	}
	
	override fun hasPermission(authentication: Authentication, targetId: Serializable, targetType: String, permission: Any): Boolean {
		val permissions: Set<String> = when(targetType) {
			"Collect" -> setOf("read", "write", "create", "delete")
			"CollectTag" -> setOf("read", "write", "create", "delete")
			"CollectCategory" -> setOf("read", "write", "create", "delete")
			"Comment" -> setOf("read", "create", "delete")
			"Notice" -> setOf("read", "create", "delete")
			else -> return false
		}
		
		if(permission !in permissions) return false
		if(permission == "create") return true
		
		val targetDomainObject = when(targetType) {
			"Collect" -> collectService.findById(targetId as Long)
			"CollectCategory" -> categoryService.findById(targetId as Long)
			"CollectTag" -> tagService.findById(targetId as Long)
			"Comment" -> commentService.findById(targetId as Long)
			"Notice" -> noticeService.findById(targetId as Long)
			else -> return false
		}
		
		val principalName: String = when(targetDomainObject) {
			is Collect -> targetDomainObject.user.username
			is CollectTag -> targetDomainObject.user.username
			is CollectCategory -> targetDomainObject.user.username
			is Comment -> targetDomainObject.sponsorByUser.username
			is Notice -> targetDomainObject.user.username
			else -> return false
		}
		
		return principalName == authentication.name
	}
}
